### ISSUE TITLE
Combination of HostBinding & new input signal api is not probably supported

### ISSUE THREAD

**Issue by Celtian** [angular/angular#53888]
### Which @angular/* package(s) are relevant/related to the feature request?

core

### Description

I am currentely trying to use new `input signal api` but unable to replace this code with DRY.

```typescript
  @HostBinding('attr.title')
  @Input({ required: true, alias: 'appModalClose' })
  public closeBtnText!: string;
```

### Proposed solution

Maybe it should be replaced by something like

```typescript
public closeBtnText = input.required<string>({ alias: 'appModalClose', host: ['attr.title'] });
```

### Alternatives considered

I was trying to find some solutions, but shit is not working

1. 

```typescript
@HostBinding('attr.title')
public closeBtnText = input.required<string>({ alias: 'appModalClose' });
```

2. Host is stacked in memory leak

```typescript
@Component({
  // eslint-disable-next-line @angular-eslint/no-host-metadata-property
  host: {
    '[title]': 'closeBtnText()',
  },
})
export class ModalCloseComponent {
  public closeBtnText = input.required<string>({ alias: 'appModalClose' });
}
```

**Comment by devversion** [comment]
My initial thought is that `@HostBinding` is expecting raw values. A `Signal` is not supported, even without signal inputs. In the `host` field of the component the signal can be read and used to e.g. bind `attr.title`. I will look more into this

**Comment by alxhub** [comment]
It's also possible to split the two, and write:

```typescript
appModalClose = input.required<string>();

@HostBinding('attr.title')
get title() { return this.appModalClose(); }
```

**Comment by devversion** [comment]
Thanks again for this issue. I've chatted a little with @pkozlowski-opensource and we both felt like supporting `@HostBinding` with signals is not necessarily something we would want to support now (without more evidence that there are requiring use-cases). Mostly because:

- We likely want to avoid encouraging two patterns. `hosts` and `@HostBinding` achieve the same thing.
- There are still ways to use `@HostBinding`, as Alex pointed out.
- Combining a decorator (`@HostBinding`) with signal inputs (that are no longer decorator-based) feels confusing. 
- Calling signals directly in the `hosts` field of e.g. `@Directive` is a more natural/obvious consumption in the reactive context. I.e. it establishes matches with the mental model that "reading signals" is necessary so that the directive/component will update whenever those change.

**Comment by Celtian** [comment]
@devversion i doubble checked

```typescript
@Directive({
  standalone: true,
  selector: 'my-dirx',
  host: { '[attr.title]': 'bla()' },
})
export class MyDir {
  bla = input.required<string>();
}
```

and it is working right now

**Comment by cyraid** [comment]
Oh, so we're stuck with using a string expression with no IDE support? That's unfortunate.

**Comment by dkimmich-onventis** [comment]
@cyraid Which IDE do you use? I am using WebStorm, and it works like a charm, with code highlighting, suggestions and everything.

![image](https://github.com/angular/angular/assets/68286501/188d0f29-8aba-47f3-ae16-b2311ba4bbb0)

**Comment by cyraid** [comment]
@dkimmich-onventis I use vscode. It would be nice if angular highlighted and syntax/error checked the host field in vscode. :)

**Comment by adammolnar** [comment]
Why is the syntax 
```
host: {
  '[class.foo]': 'bar()', 
},
```
instead of
```
host: {
  '[class.foo]': bar(), 
},
```
which would make more sense to me. In other words: why is the value a string? Just wondering.

**Comment by dkimmich-onventis** [comment]
Because `bar` in this context is not defined. `this.bar` is also not defined, as we are inside the decorator and not inside the class.

**Comment by JeanMeche** [comment]
There is no reference named `bar` in the scope the decorator.

**Comment by nickdobson** [comment]
If this is going to be the recommended approach going forward, should this style guide rule be updated?

https://angular.io/guide/styleguide#style-06-03

**Comment by devversion** [comment]
@nickdobson Yeah, we likely need to discuss this more with the team. Good point.

**Comment by c-harding** [comment]
@devversion has a function in the constructor been considered?

```ts
constructor() {
  hostBinding('class.hidden', () => isHidden());
  hostBinding('class.hidden', isHidden); // alternative: isHidden is an input
}
```

**Comment by k3nsei** [comment]
I like having hostBinding and hostListener functions more, than using not recommended by style guide host property of directive decorator.

**Comment by JeanMeche** [comment]
@k3nsei As mentionned in #54284, the style guide is know to be a bit outdated on some topics. 

As [mentionned here](https://angular.dev/guide/components/host-elements), the host property is being recommended over the usage of the hostBinding/hostListener.

**Comment by k3nsei** [comment]
@JeanMeche but HostBinding and HostListener are working with class inheritance. We are using that in our FormControl abstractions to not repeat code.That means regression in some way, because class decorators are not inherited.

**Comment by shajz** [comment]
> @k3nsei As mentionned in #54284, the style guide is know to be a bit outdated on some topics.
> 
> As [mentionned here](https://angular.dev/guide/components/host-elements), the host property is being recommended over the usage of the hostBinding/hostListener.

@JeanMeche I've noticted the recommendation on the [angular.dev website](https://angular.dev/guide/components/host-elements#the-hostbinding-and-hostlistener-decorators) is conflicting with the angular.io styleguide, as it stills recommends `HostListener/HostBinding` over the `host` property : https://angular.io/guide/styleguide#style-06-03. I was wondering why the eslint rule @angular-eslint/no-host-metadata-property disagreed with your statement. It's because it points specifically to the current styleguide https://github.com/angular-eslint/angular-eslint/blob/main/packages/eslint-plugin/docs/rules/no-host-metadata-property.md.

Should the styleguide be updated?

**Comment by k3nsei** [comment]
@shajz `angular-eslint` is community driven package, that why there could be some divergences. You can always create issue in their repository or even PR to align with latest recommendations.

**Comment by k3nsei** [comment]
@JeanMeche could this release also be synchronized with folks from `angular-eslint` as it's used by most projects using Angular?

**Comment by JeanMeche** [comment]
@k3nsei angular-eslint/angular-eslint#1772 I've suggested the change, thank you.

**Comment by Kordrad** [comment]
I like this suggestion

```
public closeBtnText = input.required<string>({ alias: 'appModalClose', hostBinding: ['attr.title'] });
```

or event something like this
![2024-05-21T07_01_39 682Z](https://github.com/angular/angular/assets/29141389/ea70c841-ee9d-450f-a6c6-b7445642fed3)
Looks like computed, but for hostBinding

**Comment by lkster** [comment]
@alxhub just wondering - is no-decorator signal-way approach even considered or you just haven't touched it yet? I'm about the way @Kordrad provided:

```ts
@Component({/*...*/})
export class Component {
    readonly customInput = input(false);
    readonly customBinding = hostBinding('attr.custom', () => this.customInput());
}
```

I'm curious about potential disadvantages of such approach if you already discussed it

**Comment by jjatczak** [comment]
https://ngxtension.netlify.app/utilities/components/host-binding/

**Comment by Kumirei** [comment]
What about hostbindings for styles? I use hostbinding to set CSS variables on the host element, but that doesn't seem to work with the host metadata property

**Comment by anthonymjones** [comment]
I would love to see this option: 

 ```
 public closeBtnText = input.required<string>({ alias: 'appModalClose', hostBinding: ['attr.title'] });
 ```

**Comment by lkster** [comment]
> I would love to see this option:
> 
> ```
> public closeBtnText = input.required<string>({ alias: 'appModalClose', hostBinding: ['attr.title'] });
> ```

I feel like this is unnecessary complication - adding too much options to just simple thing and mixing responsibilities. Potential `hostBinding` which I provided above would still be `computed`-like so this should work:

```ts

closeBtnText = input.required<string>({ alias: 'appModalClose' });

constructor() {
  hostBinding('attr.title', () => this.closeBtnText());
}
```

**Comment by Zefling** [comment]
This rule is still valid? 
https://github.com/angular-eslint/angular-eslint/blob/main/packages/eslint-plugin/docs/rules/no-host-metadata-property.md

**Comment by Memeplexx** [comment]
This could be addressed by creating an ng-host element which we could bind to more intuitively.
See https://github.com/angular/angular/issues/19119

**Comment by rhuijben** [comment]
I don't think just adding support to input signals hand	les all use cases. In my case I would like to have the value attached to something computed from multiple inputs, so more like an output signal that doesn't have to be exposed as such.

**Comment by maurotamm** [comment]
Should also add that i have also used HostBinding to modify css variables.
host: object does not allow modifying styles, so it will not work.

For example:
  `@HostBinding("style.--background-color")...`

**Comment by asherwoodcnet** [comment]
~~When I need host bindings to be inherited by child components, I don't know of a good alternative to using `@HostBinding` with a getter. For that reason in particular, it would be nice to have something like the `hostBinding("some-property", () => ...)` function. Are there any alternative suggestions for inherited bindings?~~

This comment was totally incorrect, but I'm editing rather than deleting in case anybody else stumbles across this with the same misunderstanding (let me know if I should remove instead). I did not realize the host property of a component declaration is inherited, e.g.
```ts
@Component({
  host: { "[class.parent-foo]": "true" },
})
class ParentComponent {
  @HostBinding("class.parent-bar") public parentBar = true;
}

@Component({
  host: { "[class.child-foo]": "true" },
})
class ChildComponent extends MyComponent {
  @HostBinding("class.child-bar") public childBar = true;
}
```
the child component will get `class="parent-foo parent-bar child-foo child-bar"` (I thought `parent-foo` would be omitted)

**Comment by JeanMeche** [comment]
I'll allow myself to close the issue, 
The recommendation today is to use the host `property` for host bindings and listeners. Also v20 added support for type-checking for tihs which is an additional reason to prefer them.
